### PR TITLE
Add KSPM DaemonSet reconciler

### PR DESCRIPTION
**Pull request by 0sewa0**
## Description
[DAQ-1448](https://dt-rnd.atlassian.net/browse/DAQ-1448)

Adds logic to deploy the node-collector `DaemonSet` for the KSPM feature.
(also includes some small fixes)
 

## How can this be tested?

- Image to use: `us-central1-docker.pkg.dev/cloud-platform-207208/chmu/k8s-node-config-collector`
   - The current default is not yet available 
- Use the `spec.activeGate.tlsSecretName`
   - I tested it, reason for [Fix cert handling](https://github.com/Dynatrace/dynatrace-operator/pull/3997/commits/e6f1ddca7ee12efa1e5cb0d23fe91f7949eb7b07), but double checking would be nice :D
- Use the `spec.activeGate.customProperties`
   - you have to enable the feature

Example `DynaKube`:
```
apiVersion: dynatrace.com/v1beta3
kind: DynaKube
metadata:
  name: dk
  namespace: dynatrace
spec:
  apiUrl: https://<tenant>.dev.dynatracelabs.com/api
  kspm:
    enabled: true
  activeGate:
    tlsSecretName: my-secret
    capabilities:
      - kubernetes-monitoring
    customProperties:
      value: |
        [kubernetes_monitoring]
        kubernetes_configuration_dataset_pipeline_enabled = true
        kubernetes_configuration_dataset_pipeline_include_node_config = true
  templates:
     kspmNodeConfigurationCollector:
       imageRef:
         repository: us-central1-docker.pkg.dev/cloud-platform-207208/chmu/k8s-node-config-collector
         tag: latest
```

Things to check:
- `DaemonSet` gets created
   - Related pods come up
   - Related pod respect what is configured in `templates.kspmNodeConfigurationCollector`
- `DaemonSet` gets recreated if the token-secret changes
- `Conditions` on the `DynaKube` are present and updated accordingly

[DAQ-1448]: https://dt-rnd.atlassian.net/browse/DAQ-1448?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pkg/api/v1beta3/dynakube/kspm/props.go [hunk 1]
```diff
@@ -11,3 +11,7 @@ func (kspm *Kspm) IsEnabled() bool {
 func (kspm *Kspm) GetTokenSecretName() string {
 	return kspm.name + "-" + TokenSecretKey
 }
+
+func (kspm *Kspm) GetDaemonSetName() string {
+	return kspm.name + "-" + NodeCollectorNameSuffix
```
Comment (waodim) [inline pkg/api/v1beta3/dynakube/kspm/props.go:16]: pls rewrite this as the resulting name of the ds is:

`dynakube-test--node-config-collector`

File: pkg/controllers/dynakube/kspm/daemonset/volumes.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	return mounts
+}
+
+func getTokenVolumeMount() corev1.VolumeMount {
+	return corev1.VolumeMount{
+		Name:      tokenVolumeName,
+		MountPath: tokenMountPath,
+		SubPath:   kspm.TokenSecretKey, // TODO: is this correct?
```
Comment (waodim) [inline pkg/controllers/dynakube/kspm/daemonset/volumes.go:49]: who would be the correct person to ask?

File: pkg/controllers/dynakube/kspm/daemonset/certs.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+const (
+	certVolumeName = "certs"
+	certFileName   = "ca-chain.pem"
+	certFolderPath = "/var/lib/dynatrace/ncc/customkeys"
+	certFileEnv    = "DT_CA_CERTIFICATE_FILE"
+)
+
+func needsCerts(dk dynakube.DynaKube) bool {
+	return dk.ActiveGate().TlsSecretName != ""
+}
```
Comment (albertogdd) [inline pkg/controllers/dynakube/kspm/daemonset/certs.go:19]: Move this to dk props / use existing function if there is

File: pkg/controllers/dynakube/kspm/daemonset/container.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		ReadOnlyRootFilesystem:   address.Of(true),
+		Capabilities:             &corev1.Capabilities{Drop: []corev1.Capability{"ALL"}},
+	}
+
+	return securityContext
+}
+
+func getResources(dk dynakube.DynaKube) corev1.ResourceRequirements {
+	limits := resources.NewResourceList("100m", "128Mi")
+	requests := resources.NewResourceList("100m", "128Mi")
```
Comment (albertogdd) [inline pkg/controllers/dynakube/kspm/daemonset/container.go:50]: Make these string values constants

File: pkg/controllers/dynakube/kspm/daemonset/reconciler.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	updated, err := daemonset.Query(r.client, r.apiReader, log).WithOwner(r.dk).CreateOrUpdate(ctx, ds)
+	if err != nil {
+		conditions.SetKubeApiError(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	if updated {
+		conditions.SetDaemonSetOutdated(r.dk.Conditions(), conditionType, r.dk.KSPM().GetDaemonSetName()) // needed to reset the timestamp
+		conditions.SetDaemonSetCreated(r.dk.Conditions(), conditionType, r.dk.KSPM().GetDaemonSetName())
```
Comment (albertogdd) [inline pkg/controllers/dynakube/kspm/daemonset/reconciler.go:72]: What if we create a func in `conditions` package to update the timestamp instead? This looks a bit like a gimmick

File: pkg/controllers/dynakube/kspm/daemonset/container.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+import (
+	"github.com/Dynatrace/dynatrace-operator/pkg/api/v1beta3/dynakube"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/address"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/kubeobjects/resources"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	defaultImageRepo = "public.ecr.aws/dynatrace/node-config-collector"
```
Comment (chrismuellner) [inline pkg/controllers/dynakube/kspm/daemonset/container.go:11]:
```suggestion
	defaultImageRepo = "public.ecr.aws/dynatrace/dynatrace-k8s-node-config-collector"
```

File: pkg/controllers/dynakube/kspm/daemonset/env.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+		{
+			Name:  nodeRootEnv,
+			Value: nodeRootMountPath,
+		},
+		{
+			Name:  activeGateEndpointEnv,
+			Value: getActiveGateEndpointTemplate(dk, tenantUUID),
+		},
+		{
+			Name: tokenEnv, // TODO: Remove once reading from file is supported
```
Comment (chrismuellner) [inline pkg/controllers/dynakube/kspm/daemonset/env.go:39]: reading from file should already be supported. you are currently adding the token both ways, so you should be able to remove this env var

File: pkg/controllers/dynakube/kspm/daemonset/env.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+		envs = append(envs, getCertEnv())
+	}
+
+	envs = append(envs, dk.KSPM().Env...)
+
+	return envs
+}
+
+func getActiveGateEndpointTemplate(dk dynakube.DynaKube, tenantUUID string) string {
+	activeGateEndpointTemplate := "https://%s.%s/e/%s/api/v2/kubernetes/node-config"
```
[the remark this answers is not in the document]
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/env.go:61]: I just followed what was in the ticket 😅 
can try without it

File: pkg/controllers/dynakube/kspm/daemonset/reconciler.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		return nil, err
+	}
+
+	return ds, nil
+}
+
+func (r *Reconciler) getUpdateStrategy() appsv1.DaemonSetUpdateStrategy {
+	updateStrategy := r.dk.KSPM().UpdateStrategy
+
+	if updateStrategy.Type == "" {
```
[the remark this answers is not in the document]
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/reconciler.go:115]: this was my attempt to do defaults without replacing the updateStrategy with a pointer, but I think I will just do that

File: pkg/controllers/dynakube/kspm/daemonset/env.go (hunk 7 above)
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/env.go:39]: tried it, with the image I have, and I get an error:
```
Failed contacting ActiveGate: not supported empty Token
```

Either I put the token in the wrong place, or my image doesn't support it yet

File: pkg/controllers/dynakube/kspm/daemonset/env.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/env.go:61]: I just removed it (I dunno how correct my URL is now), but didn't work
```
2024/10/29 18:29:09 DT_ACTIVEGATE_ENDPOINT environment variable has issues: invalid url received for validation: "https://zib50933zib50933zib50933zib50933zib5093-activegate.dynatrace/e/api/v2/kubernetes/node-config"
```

File: pkg/controllers/dynakube/kspm/daemonset/env.go (hunk 7 above)
Comment (chrismuellner) [reply pkg/controllers/dynakube/kspm/daemonset/env.go:39]: we are still using the old dev image, so we should try to update that to a newer one. I can help with that if you want

File: pkg/controllers/dynakube/kspm/daemonset/env.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chrismuellner) [reply pkg/controllers/dynakube/kspm/daemonset/env.go:61]: did you try without `/e/`? it seems extension is already using a similar entrypoint: https://github.com/Dynatrace/dynatrace-operator/blob/a22de1b1756719a4051a61042851f2b028e16b0e/pkg/controllers/dynakube/activegate/capability/capability.go#L190

[the remark this answers is not in the document]
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/env.go:61]:
```
2024/10/30 08:16:01 DT_ACTIVEGATE_ENDPOINT environment variable has issues: invalid url received for validation: "https://zib50933zib50933zib50933zib50933zib5093-activegate.dynatrace/api/v2/kubernetes/node-config"
```

with port:
```
2024/10/30 08:25:14 DT_ACTIVEGATE_ENDPOINT environment variable has issues: invalid url received for validation: "https://zib50933zib50933zib50933zib50933zib5093-activegate.dynatrace:443/api/v2/kubernetes/node-config"
```

File: pkg/api/v1beta3/dynakube/kspm/props.go (hunk 1 above)
Comment (0sewa0) [reply pkg/api/v1beta3/dynakube/kspm/props.go:16]: fixed

File: pkg/controllers/dynakube/kspm/daemonset/volumes.go (hunk 2 above)
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/volumes.go]: tested it, works 😄

File: pkg/controllers/dynakube/kspm/daemonset/container.go (hunk 4 above)
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/container.go]: done

File: pkg/controllers/dynakube/kspm/daemonset/reconciler.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+		}
+
+		query := daemonset.Query(r.client, r.apiReader, log)
+		err := query.Delete(ctx, &appsv1.DaemonSet{ObjectMeta: metav1.ObjectMeta{Name: r.dk.KSPM().GetDaemonSetName(), Namespace: r.dk.Namespace}})
+
+		if err != nil {
+			log.Error(err, "failed to clean-up KSPM config-secret")
+		}
+
+		meta.RemoveStatusCondition(r.dk.Conditions(), conditionType)
```
[the remark this answers is not in the document]
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/reconciler.go]: done

File: pkg/controllers/dynakube/kspm/daemonset/reconciler.go (hunk 5 above)
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/reconciler.go:72]: it's fine as is now, we can think about it in general how we want to handle this (will create a follow up)

File: pkg/controllers/dynakube/kspm/daemonset/container.go (hunk 6 above)
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/container.go]: done

File: pkg/controllers/dynakube/kspm/daemonset/env.go (hunk 7 above)
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/env.go]: done

File: pkg/controllers/dynakube/kspm/daemonset/env.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (0sewa0) [reply pkg/controllers/dynakube/kspm/daemonset/env.go:49]: leaving it as is for now

File: pkg/controllers/dynakube/kspm/daemonset/reconciler.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+			return nil // no condition == nothing is there to clean up
+		}
+
+		defer meta.RemoveStatusCondition(r.dk.Conditions(), conditionType)
+
+		query := daemonset.Query(r.client, r.apiReader, log)
+		err := query.Delete(ctx, &appsv1.DaemonSet{ObjectMeta: metav1.ObjectMeta{Name: r.dk.KSPM().GetDaemonSetName(), Namespace: r.dk.Namespace}})
+
+		if err != nil {
+			log.Error(err, "failed to clean-up KSPM config-secret")
```
[the remark this answers is not in the document]
Comment (waodim) [reply pkg/controllers/dynakube/kspm/daemonset/reconciler.go:52]: If I am not mistaken, here we handle the daemonset and not a secret, don't we?

**Review by waodim — APPROVED** [review]

**Review by albertogdd — APPROVED** [review]
LGTM, although I keep getting this error on node collector pod logs.
```
2024/10/31 09:39:40.498276 Error: Failed contacting ActiveGate: failed to perform request: Get "https://dynakube-activegate.dynatrace/e/dib18963/api/v2/kubernetes/node-config/status": dial tcp: lookup dynakube-activegate.dynatrace on 10.100.0.10:53: no such host, url: "https://dynakube-activegate.dynatrace/e/dib18963/api/v2/kubernetes/node-config/status"
2024/10/31 09:39:40.498300 Info: Collection not executed because of previous error. Retrying in 60 seconds.
```

**Pull request merged**